--- openssl-1.0.1e/crypto/objects/objects.txt	2013-02-11 19:26:04.000000000 +0400
+++ openssl-1.0.1e/crypto/objects/objects.txt	2014-03-27 16:09:49.949211881 +0400
@@ -775,6 +775,9 @@
 !Cname no-rev-avail
 id-ce 56		: noRevAvail		: X509v3 No Revocation Available
 
+!Cname selinux-context
+id-ce 57		: selinuxContext	: X509v3 Selinux Context Extension
+
 # From RFC5280
 ext-key-usage 0		: anyExtendedKeyUsage	: Any Extended Key Usage
 
--- openssl-1.0.1e/crypto/objects/objects.h	2013-04-30 17:53:18.000000000 +0400
+++ openssl-1.0.1e/crypto/objects/objects.h	2014-03-27 16:17:36.905424061 +0400
@@ -501,6 +501,11 @@
 #define NID_authority_key_identifier	90
 #define OBJ_authority_key_identifier	OBJ_id_ce,35L
 
+#define SN_selinux_context		"selinuxContext"
+#define LN_selinux_context		"X509v3 Selinux Context Extension"
+#define NID_selinux_context		920
+#define OBJ_selinux_context		OBJ_id_ce,57L
+
 #define SN_bf_cbc			"BF-CBC"
 #define LN_bf_cbc			"bf-cbc"
 #define NID_bf_cbc			91
